core/src/main/java/org/apache/iceberg/TableMetadataParser.java

```diff
@@ -380,7 +380,12 @@ static TableMetadata fromJson(FileIO io, String metadataLocation, JsonNode node)
       // parse the spec array
       ImmutableList.Builder<PartitionSpec> builder = ImmutableList.builder();
       for (JsonNode spec : specArray) {
-        builder.add(PartitionSpecParser.fromJson(schema, spec));
+        UnboundPartitionSpec unboundSpec = PartitionSpecParser.fromJson(spec);
+        if (unboundSpec.specId() == defaultSpecId) {
+          builder.add(unboundSpec.bind(schema));
```

Comment:
@singhpk234 Can you elaborate a bit more? I'm not sure that I understand the issue.

For V1, using the `spec` instead of `specs`, nothing changes. Since it takes the other branch: https://github.com/apache/iceberg/blob/24d5a53a0f90b47203eb3b93b0d141e154c36abe/core/src/main/java/org/apache/iceberg/TableMetadataParser.java#L393-L402

Reply:
Apologies for not being clear, 
What I meant here was, mapping the current specId to the current spec might still cause issue as in case of table of V1 format, (Considering the UT in this PR, last DDL) The current spec will have a void transform for `day_of_ts` since `day_of_ts` is dropped from partitioning (in V1 format, we replace this transform with a void transform). Now when we attempt to bind current schema with current partition spec, current schema will not have `day_of_ts`, but in `PartitionSpec#checkCompatibility` we will try to find `schema.findType(field.sourceId())`, from current schema for void transform of `day_of_ts` and it will fail, as we have removed `day_of_ts` from current schema.


A sample UT for repro (modified from this PR) : 

```java
  @Test
  public void testDropColumnOfOldPartitionField() {
    // default table created in v1 format
    sql(
        "CREATE TABLE %s (id bigint NOT NULL, ts timestamp, day_of_ts date) USING iceberg PARTITIONED BY (day_of_ts)",
        tableName);

    sql("ALTER TABLE %s REPLACE PARTITION FIELD day_of_ts WITH days(ts)", tableName);
    sql("ALTER TABLE %s DROP COLUMN day_of_ts", tableName);
  }
```

Reply:
@singhpk234 Thanks for the thorough explanation. I wasn't aware of the replacement by a void transform, and it indeed has a similar issue:

![image](https://user-images.githubusercontent.com/1134248/188821063-aaaebf03-213f-4ba4-9103-83161cbbb332.png)

Reply:
@Fokko , This [explanation](https://github.com/apache/iceberg/pull/924#issue-600524755) from @rdblue, nicely explains the motivation behind introducing void transforms in v1 format.

Reply:
Thanks for linking @singhpk234, and that is indeed a very clear explanation and it makes a lot of sense.

I've added a condition to skip the validation when we encounter a `VoidTransform`. Since it is almost always compatible, I think we should be okay with that, but curious to hear what others think.